### PR TITLE
Add RequestLogEvent emitters config to graphite-emitter

### PR DESCRIPTION
**Pull request by jon-wei**
If using the graphite-emitter and `druid.request.logging.type=emitter`, the graphite-emitter currently generates an error for every request log event:

```
var/sv/broker.log:2017-08-09T22:39:16,806 ERROR [qtp2067233363-77] io.druid.emitter.graphite.GraphiteEmitter - unknown event type [class io.druid.server.log.EmittingRequestLogger$RequestLogEvent]
var/sv/broker.log:2017-08-09T22:39:17,023 ERROR [qtp2067233363-71] io.druid.emitter.graphite.GraphiteEmitter - unknown event type [class io.druid.server.log.EmittingRequestLogger$RequestLogEvent]
```

This PR allows the graphite-emitter to optionally forward RequestLogEvents to other emitters instead of generating an error message.

### CODE REVIEW

**Comment by jon-wei** [comment]
@b-slim Can you please take a look at this PR?

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitter.java [hunk 1]
```diff
@@ -119,8 +123,12 @@ public void emit(Event event)
         log.error(e, "got interrupted with message [%s]", e.getMessage());
         Thread.currentThread().interrupt();
       }
-    } else if (!emitterList.isEmpty() && event instanceof AlertEvent) {
-      for (Emitter emitter : emitterList) {
+    } else if (event instanceof EmittingRequestLogger.RequestLogEvent) {
+      for (Emitter emitter : requestLogEmitters) {
```
Comment (b-slim) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitter.java]: are we dropping the event if the list is null or empty ?

Comment (jon-wei) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitter.java]: yes, they'll be dropped if the user didn't configure any forwarding emitters

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java [hunk 2]
```diff
@@ -68,6 +68,18 @@ public Emitter apply(String s)
...
-    return new GraphiteEmitter(graphiteEmitterConfig, emitters);
+
+    List<Emitter> requestLogEmitters = Lists.transform(
+        graphiteEmitterConfig.getRequestLogEmitters(),
+        new Function<String, Emitter>()
+        {
+          @Override
+          public Emitter apply(String s)
+          {
+            return injector.getInstance(Key.get(Emitter.class, Names.named(s)));
```
Comment (drcrallen) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java]: This has the potential to fail very late. Suggest eagerly creating the list with `ImmutableList.copyOf`

Comment (jon-wei) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java]: @drcrallen Added ImmutableList.copyOf

**Comment by b-slim** [comment]
```
Failed tests: 

  DruidCoordinatorBalancerTest.testMoveToEmptyServerBalancer:185
```

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java [hunk 3]
```diff
@@ -108,6 +116,7 @@ public int hashCode()
     result = 31 * result + getMaxQueueSize().hashCode();
     result = 31 * result + getDruidToGraphiteEventConverter().hashCode();
     result = 31 * result + (getAlertEmitters() != null ? getAlertEmitters().hashCode() : 0);
+    result = 31 * result + (getRequestLogEmitters() != null ? getRequestLogEmitters().hashCode() : 0);
```
Comment (leventov) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java]: Could use `Objects.hash()` in this method

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java [hunk 4]
```diff
@@ -91,6 +94,11 @@ public boolean equals(Object o)
         : that.getAlertEmitters() != null) {
       return false;
     }
+    if (getRequestLogEmitters() != null
```
Comment (leventov) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java]: Could use `Objects.equals()`

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java [hunk 5]
```diff
@@ -57,17 +58,33 @@ public void configure(Binder binder)
...
-        graphiteEmitterConfig.getAlertEmitters(),
-        new Function<String, Emitter>()
-        {
-          @Override
-          public Emitter apply(String s)
-          {
-            return injector.getInstance(Key.get(Emitter.class, Names.named(s)));
-          }
-        }
+    List<Emitter> emitters = ImmutableList.copyOf(
```
Comment (leventov) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java]: Please use stream API and lambda

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java [hunk 6]
```diff
@@ -57,17 +58,33 @@ public void configure(Binder binder)
...
+              public Emitter apply(String s)
+              {
+                return injector.getInstance(Key.get(Emitter.class, Names.named(s)));
+              }
+            }
+        )
     );
-    return new GraphiteEmitter(graphiteEmitterConfig, emitters);
+
+    List<Emitter> requestLogEmitters = ImmutableList.copyOf(
```
Comment (leventov) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java]: Same

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java [hunk 7]
```diff
@@ -50,6 +50,9 @@
   final private List<String> alertEmitters;
 
   @JsonProperty
+  final private List<String> requestLogEmitters;
```
Comment (drcrallen) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java]: Is there a way to set this to `null`? it seems there are a few spots in the code that will break if this ends up being `null`

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java [hunk 8]
```diff
@@ -122,13 +131,15 @@ public GraphiteEmitterConfig(
...
       @JsonProperty("alertEmitters") List<String> alertEmitters,
+      @JsonProperty("requestLogEmitters") List<String> requestLogEmitters,
       @JsonProperty("emitWaitTime") Long emitWaitTime,
       @JsonProperty("waitForEventTime") Long waitForEventTime
   )
   {
     this.waitForEventTime = waitForEventTime == null ? DEFAULT_GET_TIMEOUT : waitForEventTime;
     this.emitWaitTime = emitWaitTime == null ? 0 : emitWaitTime;
     this.alertEmitters = alertEmitters == null ? Collections.<String>emptyList() : alertEmitters;
+    this.requestLogEmitters = requestLogEmitters == null ? Collections.<String>emptyList() : requestLogEmitters;
```
Comment (drcrallen) [inline extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterConfig.java]: ah ok, this should protect it

**Comment by b-slim** [comment]
@jon-wei can you please solve the conflicts this PR is good to go

File: extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java (hunk 5 above)
Comment (jon-wei) [reply extensions-contrib/graphite-emitter/src/main/java/io/druid/emitter/graphite/GraphiteEmitterModule.java]: changed these to use lambdas

**Comment by jon-wei** [comment]
merged with master, thanks for the review!

**Pull request merged**